cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ImpersonationHandler.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.gateway.handlers;
+
+import co.cask.cdap.common.BadRequestException;
+import co.cask.cdap.data2.security.ImpersonationInfo;
+import co.cask.cdap.data2.security.ImpersonationUtils;
+import co.cask.cdap.security.DefaultUGIProvider;
+import co.cask.cdap.security.TokenSecureStoreUpdater;
+import co.cask.http.AbstractHttpHandler;
+import co.cask.http.HttpResponder;
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.twill.api.SecureStore;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.jboss.netty.handler.codec.http.HttpRequest;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.util.concurrent.Callable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * Resolves the UGI for a given namespace, acquires the delegation tokens for that UGI,
+ * using {@link TokenSecureStoreUpdater}, and serializes these Credentials to a location.
+ *
+ * Response with the location to which the credentials were serialized to, as well as the UGI's short username
+ */
+// we don't share the same version as other handlers, so we can upgrade/iterate faster
+@Path("/v1")
+public class ImpersonationHandler extends AbstractHttpHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultUGIProvider.class);
+  private static final Gson GSON = new Gson();
+
+  private final DefaultUGIProvider ugiProvider;
+  private final TokenSecureStoreUpdater tokenSecureStoreUpdater;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public ImpersonationHandler(DefaultUGIProvider tokenRequester,
```

Comment:
Why takes `DefaultUGIProvider` instead of `UGIProvider`. In general we should use the binding at the interface level, not the concrete class.